runtime/old_parser/type.go

```diff
@@ -0,0 +1,988 @@
+				"expected identifier after %s, got %s",
+				lexer.TokenDot,
+				nestedToken.Type,
+			)
+		}
+
+		nestedIdentifier := p.tokenToIdentifier(nestedToken)
+
+		// Skip the identifier
+		p.next()
+
+		nestedIdentifiers = append(
+			nestedIdentifiers,
+			nestedIdentifier,
+		)
+
+	}
+
+	return ast.NewNominalType(
+		p.memoryGauge,
+		p.tokenToIdentifier(token),
+		nestedIdentifiers,
+	), nil
+}
+
+func defineArrayType() {
+	setTypeNullDenotation(
+		lexer.TokenBracketOpen,
+		func(p *parser, startToken lexer.Token) (ast.Type, error) {
+
+			elementType, err := parseType(p, lowestBindingPower)
+			if err != nil {
+				return nil, err
+			}
+
+			p.skipSpaceAndComments()
+
+			var size *ast.IntegerExpression
+
+			if p.current.Is(lexer.TokenSemicolon) {
+				// Skip the semicolon
+				p.nextSemanticToken()
+
+				if !p.current.Type.IsIntegerLiteral() {
+					p.reportSyntaxError("expected positive integer size for constant sized type")
+
+					// Skip the invalid non-integer literal token
+					p.next()
+
+				} else {
+					numberExpression, err := parseExpression(p, lowestBindingPower)
+					if err != nil {
+						return nil, err
+					}
+
+					integerExpression, ok := numberExpression.(*ast.IntegerExpression)
+					if !ok || integerExpression.Value.Sign() < 0 {
+						p.reportSyntaxError("expected positive integer size for constant sized type")
+					} else {
+						size = integerExpression
+					}
+				}
+			}
+
+			p.skipSpaceAndComments()
+
+			endToken, err := p.mustOne(lexer.TokenBracketClose)
+			if err != nil {
+				return nil, err
+			}
+
+			typeRange := ast.NewRange(
+				p.memoryGauge,
+				startToken.StartPos,
+				endToken.EndPos,
+			)
+
+			if size != nil {
+				return ast.NewConstantSizedType(
+					p.memoryGauge,
+					elementType,
+					size,
+					typeRange,
+				), nil
+			} else {
+				return ast.NewVariableSizedType(
+					p.memoryGauge,
+					elementType,
+					typeRange,
+				), nil
+			}
+		},
+	)
+}
+
+func defineOptionalType() {
+	defineType(postfixType{
+		tokenType:    lexer.TokenQuestionMark,
+		bindingPower: typeLeftBindingPowerOptional,
+		leftDenotation: func(p *parser, left ast.Type, tokenRange ast.Range) ast.Type {
+			return ast.NewOptionalType(
+				p.memoryGauge,
+				left,
+				tokenRange.EndPos,
+			)
+		},
+	})
+
+	defineType(postfixType{
+		tokenType:    lexer.TokenDoubleQuestionMark,
+		bindingPower: typeLeftBindingPowerOptional,
+		leftDenotation: func(p *parser, left ast.Type, tokenRange ast.Range) ast.Type {
+			return ast.NewOptionalType(
+				p.memoryGauge,
+				ast.NewOptionalType(
+					p.memoryGauge,
+					left,
+					tokenRange.StartPos,
+				),
+				tokenRange.EndPos,
+			)
+		},
+	})
+}
+
+func defineReferenceType() {
+	defineType(prefixType{
+		tokenType:    lexer.TokenAmpersand,
+		bindingPower: typeLeftBindingPowerReference,
+		nullDenotation: func(p *parser, right ast.Type, tokenRange ast.Range) ast.Type {
+			return ast.NewReferenceType(
+				p.memoryGauge,
+				nil,
```

Comment:
Reference types parsed by the old parser are always created with a `nil` authorization. This should indicate to whatever code we use this for that the reference type was created in an older version of Cadence.

Reply:
Does it include the `IsAuthorized` flag?

Reply:
No, the 1.0 Cadence AST doesn't have support for the `IsAuthorized` flag, and I don't really think we need to add it; whether or not a reference was authorized in older versions of Cadence is irrelevant for their semantics w.r.t. migrations, so I don't think it's necessary for us to care about that information when doing upgrade checks.

Reply:
Might be worth making this package independent of assumptions on how it will be used, and add it to the AST as a "legacy" field and then users of the package can decide to use it or not

Reply:
Added this in [110bd77](https://github.com/onflow/cadence/pull/3039/commits/110bd77514aaab13d50cb9eab935e85c026b4da7)